### PR TITLE
Option to add custom options on subscription form.

### PR DESCRIPTION
**Pull request by ghost**
Custom fields are added in admin menu with format key|value.
Custom options will be shown on the registration form as checkboxes.

### CODE REVIEW

**Review by kaittodesk — CHANGES_REQUESTED** [review]
Can you build some sort of validation logic for AdminForm `smaily_custom_fields` field in order to verify if user has inserted a value that can be used SubscriberForm in block? If the structure of the value does not conform with the structure SubscriberForm block is trying to parse, it could start spewing warnings and errors.

File: smaily_for_drupal/src/Form/AdminForm.php [hunk 1]
```diff
@@ -86,13 +86,41 @@ public function buildForm(array $form, FormStateInterface $form_state) {
...
 
+  /**
+   * {@inheritdoc}
+   */
+  public function validateForm(array &$form, FormStateInterface $form_state) {
+    $lines = $form_state->getValue('customfields');
+    if (!empty($lines)) {
+      foreach (explode("\n", $lines) as $line) {
+        // Match any non-whitespace character with "|" and with any non-whitespace character.
+        if (!preg_match('/[\S+]\|[\S+]/', $line)) {
```
Comment (kaittodesk) [inline smaily_for_drupal/src/Form/AdminForm.php]: This regular expression will allow value `foo|bar|Baz` through. Where field name will be `foo|bar` and value `Baz`, but it shouldn't be allowed to save at all. Also user might have this kind of value there also - `foo|My favourite character is pipe (|)` - safest option would be to not allow user to save this value.

**Review by kaittodesk — CHANGES_REQUESTED** [review]

File: smaily_for_drupal/src/Form/AdminForm.php (hunk 1 above)
Comment (ghost) [reply smaily_for_drupal/src/Form/AdminForm.php]: Updated the pattern. Is it sufficient?

Comment (kaittodesk) [reply smaily_for_drupal/src/Form/AdminForm.php]: Yep.

**Review by kaittodesk — APPROVED** [review]

**Pull request merged**